### PR TITLE
Add mainFilesContainingFile and IndexDelegate

### PR DESCRIPTION
**Pull request by benlangmuir**
Expose the mainFile query to lookup the main files that include a given file, which will be used by sourcekit-lsp when looking up header build settings. Also, expose the index delegate processing events so that a client can track ongoing work and see when the index finishes updating.

### CODE REVIEW

**Comment by benlangmuir** [comment]
@swift-ci please test

File: Sources/ISDBTestSupport/ForwardingIndexDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
```
Comment (akyrtzi) [inline Sources/ISDBTestSupport/ForwardingIndexDelegate.swift]: `2020` for new files 😉

Comment (benlangmuir) [reply Sources/ISDBTestSupport/ForwardingIndexDelegate.swift]: I did the indexstore-db side of this in December. I just sat on the changes until I could test them thoroughly from the sourcekit-lsp side.  I can still update the dates if you think it's correct :shrug:  Wish we could just remove the dates from the headers.  No one is consistent about keeping them updated in existing files.

Comment (akyrtzi) [reply Sources/ISDBTestSupport/ForwardingIndexDelegate.swift]: It's fine, I don't really care much.

**Pull request merged**